resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala

```diff
@@ -793,6 +793,25 @@ private[spark] class Client(
       // distributed file.
       amKeytabFileName.foreach { kt => props.setProperty(KEYTAB.key, kt) }
 
+      // Upload user provided ivysettings.xml file to the distributed cache
+      val ivySettings = sparkConf.getOption("spark.jars.ivySettings")
+      if (isClusterMode && ivySettings.isDefined) {
+        val ivySettingsFile = new File(ivySettings.get)
+        require(ivySettingsFile.exists(), s"Ivy settings file $ivySettingsFile not found")
+        require(ivySettingsFile.isFile(),
+          s"Ivy settings file $ivySettingsFile is not a normal file")
+        // Generate a file name that can be used for the ivySettings file, that does not conflict
+        // with any other conf file.
+        val amIvySettingsFileName = ivySettingsFile.getName() + "-" + UUID.randomUUID().toString
+        confStream.putNextEntry(new ZipEntry(amIvySettingsFileName))
```

Comment:
any reason not to just use distribute() rather then putting into the conf files?

Reply:
This would help save some RPCs in an HDFS environment, more context in 
https://github.com/apache/spark/pull/31591#pullrequestreview-594367541 See drawback (2)

Reply:
I think also, semantically, this is more similar to a Spark configuration file than a user-provided file and thus makes sense to have within the conf zip as opposed to living with the user JARs/files. This is similar to the approach used for `log4j.properties` and `metrics.properties` above.

Reply:
I get that its kind of like a conf file but its not, there is no template file in the conf directory for this, it's a user specified config and not automatically picked up, so I would rather keep things consistent and have it treated like all the other configs similar to it.
Also we don't want to support this being in HDFS?   I think that comes for free if you just use distribute, I don't see a few rpc calls as being a big deal here especially since it adds functionality. I guess if you allow that you might need more code on the client side to download it.  Its a bit unfortunately we are not consistent with these things.  

Personally I would prefer us to stay consistent and this act just like other files users can specify.  it should allow being in HDFS and it should be distributed via distribute() just like all the other files users can specify.

Reply:
> I get that its kind of like a conf file but its not, there is no template file in the conf directory for this, it's a user specified config and not automatically picked up ... Its a bit unfortunately we are not consistent with these things.

The same is (mostly) true for examples like `metrics.properties`, which is an external/non-Spark file, the only difference being that it is automatically picked up vs. being requested by a user. I don't have much opinion on whether or not this is how it should be -- I agree with you that more clear guidelines/consistency in this area would be nice.

I don't have too much opinion between storing in the conf object and leveraging `distribute`, I do agree that we should support remote file systems (e.g. HDFS) so this is a good point. @HyukjinKwon do you have any opinion here since you participated in some of the earlier conversations?

Reply:
I have updated the PR using `distribute()` to localize the ivySettings file instead of putting the file in Spark conf zip. I also had to slightly modify the ivySettings loading code in `SparkSubmit.scala` since it did not support local URIs in the first place. I also added tests for local and non-local URIs for ivySettings.